tests/queries/0_stateless/02952_simple_streaming_read.sh

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+opts=(
+    "--allow_experimental_analyzer=1"
+    "--allow_experimental_streaming=1"
+)
+
+function run_test {
+    printf "=====\n"
+    printf "running test for engine: %s\n" "$1"
+    sleep 0.5
+
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "DROP TABLE IF EXISTS t_streaming_test"
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "CREATE TABLE t_streaming_test (a String, b UInt64) ENGINE = $1"
+
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT count() FROM t_streaming_test"
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test VALUES ('a', 100)"
+    sleep 0.5
+
+    # spawn stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a, a || '-' || a, b * b FROM t_streaming_test STREAM" &
+    sleep 0.5
+
+    # insert some data into stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test (*) select number as a, number as b from numbers(7)"
+    sleep 0.5
+
+    # stop reading by killing client job
+    pkill -P $$
+    sleep 0.5
+
+    # clear table
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "TRUNCATE t_streaming_test"
+
+    # spawn 2 streams
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a FROM t_streaming_test STREAM WHERE b % 2 == 0" &
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "SELECT a FROM t_streaming_test STREAM WHERE b % 2 == 1" &
+    sleep 0.5
+
+    # insert some data into stream
+    $CLICKHOUSE_CLIENT "${opts[@]}" -q "INSERT INTO t_streaming_test (*) select 'same-x-2' as a, number as b from numbers(2)"
+    sleep 0.5
+
+    # stop reading by killing client job
+    pkill -P $$
+    sleep 0.5
+}
+
+for engine_type in 'MergeTree() ORDER BY (a)' 'TinyLog' 'StripeLog' 'Memory';
+do
+   run_test "$engine_type"
+done
+
+
```

Comment:
Thanks for this advice, by default many of this queries will not emit anything, but they will not stuck - at any time you can cancel the query. For now i can add some check that will fail queries with union or CreatingSet step from streaming source, but in the future it can be possible to replace such query plan steps with streaming analogs.

Second query for example is a good case for supportsStreaming IStorage method

Reply:
Right, I guess it makes sense to just say that STREAM makes the query run forever, and there's no need to prevent the user from trying to use it where it doesn't make sense (e.g. in IN subquery - the user is maybe supposed to understand that the main query can't make progress until the IN subquery completes).

So the expected behavior for all of the above queries is to either stall or work. May make sense to smoke test it anyway, to make sure it doesn't crash, if it's easy to do (just start the query, sleep 0.5, kill it, repeat for a bunch of different queries?). For UNION ALL maybe also check that it actually works, i.e. sees inserts from both tables?

Reply:
added one more test to check the corner cases